sdk/daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/crypto/NodeHashSpec.scala

```diff
@@ -0,0 +1,893 @@
+      Hash.hashNode(node, subNodes)
+    }
+
+    "be stable" in {
+      hashRollbackNode(rollbackNode) shouldBe defaultHash
+    }
+
+    "throw if some nodes are missing" in {
+      an[IncompleteTransactionTree] shouldBe thrownBy {
+        Hash.hashNode(rollbackNode)
+      }
+    }
+
+    "not hash NodeIds" in {
+      Hash.hashNode(
+        rollbackNode
+          // Change the node Ids values but not the nodes
+          .copy(children = rollbackNode.children.map(nodeId => NodeId(nodeId.index + 1))),
+        subNodes.map { case (nodeId, node) =>
+          NodeId(nodeId.index + 1) -> node
+        },
+      ) shouldBe defaultHash
+    }
+
+    "not produce collision in children" in {
+      hashRollbackNode(
+        rollbackNode.copy(
+          children = rollbackNode.children.reverse
+        )
+      ) should !==(defaultHash)
+    }
+  }
+
+  "TransactionBuilder" should {
+    val roots = ImmArray(NodeId(2), NodeId(5))
+    val nodes = Map(
+      NodeId(0) -> createNode,
+      NodeId(1) -> fetchNode,
+      NodeId(2) -> exerciseNode,
+      NodeId(3) -> lookupNode,
+      NodeId(4) -> lookupNode2,
+      NodeId(5) -> rollbackNode,
+    )
+    val transaction = VersionedTransaction(
+      version = LanguageVersion.v2_1,
+      roots = roots,
+      nodes = nodes,
+    )
+
+    val defaultHash = Hash
+      .fromString("43cbd72afc15db4c35f88a3abcc5812f228b403ef6d9a312dd8e0bce564d54aa")
+      .getOrElse(fail("Invalid hash"))
+
+    "be stable" in {
+      Hash.hashTransaction(transaction) shouldBe defaultHash
+    }
+
+    "throw if some nodes are missing" in {
+      an[IncompleteTransactionTree] shouldBe thrownBy {
+        Hash.hashTransaction(
+          VersionedTransaction(
+            version = LanguageVersion.v2_1,
+            roots = roots,
+            nodes = nodes.drop(2),
+          )
+        )
+      }
+    }
+
+    "not hash NodeIds" in {
+      Hash.hashTransaction(
+        VersionedTransaction(
+          version = LanguageVersion.v2_1,
+          roots = ImmArray(NodeId(8), NodeId(44)),
+          nodes = Map(
+            NodeId(75) -> createNode,
+            NodeId(84) -> fetchNode,
+            NodeId(8) -> exerciseNode.copy(children = ImmArray(NodeId(75), NodeId(84))),
+            NodeId(15) -> lookupNode,
+            NodeId(2009) -> lookupNode2,
+            NodeId(44) -> rollbackNode.copy(children = ImmArray(NodeId(15), NodeId(2009))),
+          ),
+        )
+      ) shouldBe defaultHash
+    }
+
+    "not produce collision in children" in {
+      Hash.hashTransaction(
+        VersionedTransaction(
+          version = LanguageVersion.v2_1,
+          roots = roots.reverse,
+          nodes = nodes,
+        )
+      ) should !==(defaultHash)
+    }
+
+    "produce a debug output" in {
+      val os = new DebugStringOutputStream
+      val hash = Hash.hashTransactionDebug(transaction, os)
+      hash shouldBe defaultHash
+
+      os.result shouldBe """_Transaction_
```

Comment:
Maybe ignore this, was more to show what the output is like. I'd rather add per-value examples and one for each node instead of this big one